### DISCUSSION TITLE
How to separate development environment and production environment?

### DISCUSSION THREAD

**Discussion by SkullCutter02** [directus/directus#11612] [Archive]
Back when I was using Strapi, it really easy to have a local copy of the CMS on my localhost, and a deployed version on somewhere on Heroku where the data model editor is disabled and uses a production database. While playing around with Directus, I noticed Directus saves everything on the database, and there's no schema javascript files. So I tried deploying my local copy of Directus, and sure enough none of the models have been saved on Heroku. So now if I want to develop directus, I need to use the production version, as if I have a localhost version it won't sync with the production one. 

I was wondering if there's a mode that can emulate what Strapi does? Meaning I can have a localhost copy that I use to develop on, then I can deploy the local copy to Heroku, and this production copy will have the exact same models as the local copy, but uses a production postgres database from RDS and has disabled data model editing, and is only used for content editing?

**Comment by xu4wang** [comment]
I had same question a few days ago. seems the cli command snapshot can be used to achieve this... 

I do not test it in real production env yet.

**Reply by SkullCutter02** [reply]
May you elaborate? What do you mean by a cli command snapshot?

**Reply by xu4wang** [reply]
It's documented here  
https://docs.directus.io/reference/cli/#snapshot-the-data-model

I tried it by:
1. Take a snapshot of current database
2. make some modifications to database scheme
3. restore the update database to original one ( mimic the production environment without the modifications in above step).
4. Apply the snapshot we took from step 1.
5. Check the scheme was actually updated in the database, also the data were not lost.

```
#make a snapshot of current scheme

MacBook-Pro-2:api wangxu$ npx directus schema snapshot ./snapshot_old.yaml
14:05:11 ✨ Snapshot saved to /Users/wangxu/Documents/github/directus/api/snapshot_old.yaml

#create a new collection named articles via the app
...

# make another snapshot
MacBook-Pro-2:api wangxu$ npx directus schema snapshot ./snapshot_new.yaml
14:06:44 ✨ Snapshot saved to /Users/wangxu/Documents/github/directus/api/snapshot_new.yaml

#check the difference

MacBook-Pro-2:api wangxu$ diff snapshot_old.yaml snapshot_new.yaml
526a527,556
>   - collection: articles
>     meta:
>       collection: articles
>       icon: null
>       note: null
>       display_template: null
>       hidden: false
>       singleton: false
>       translations: null
>       archive_field: status
>       archive_app_filter: true
......
....

# remove the database, and restore using an old copy.. this is sqlite.
MacBook-Pro-2:api wangxu$ mv data.db data_old.db
MacBook-Pro-2:api wangxu$ cp /path/to/another/backup/data.db ./

MacBook-Pro-2:api wangxu$ npx directus schema apply ./snapshot_new.yaml
? The following changes will be applied:

Collections:
  - Create articles

Fields:
  - Update Pages.id
    - Set hidden to false
  - Create articles.id
  - Create articles.status
  - Create articles.sort
  - Create articles.user_created
  - Create articles.date_created
  - Create articles.user_updated
  - Create articles.date_updated

Relations:
  - Create articles.user_updated-> directus_users
  - Create articles.user_created-> directus_users

Would you like to continue? Yes
14:09:21 ✨ Snapshot applied successfully
MacBook-Pro-2:api wangxu$
```

**Reply by SkullCutter02** [reply]
Thanks a lot! I tried using it and it worked! Had a development version on my local machine, and deployed directus onto heroku then run the snapshot update command! Now I'm wondering if there's sort of an ENV variable I can toggle that disables the data model editor? This option would be useful for the production version of the app.

**Reply by xu4wang** [reply]
You can create new role for public access you production env.

The data model editor is disabled by default for non-admin role.

**Reply by jiri-matejka** [reply]
Linked changed to https://docs.directus.io/self-hosted/cli.html#migrate-schema-to-a-different-environment
